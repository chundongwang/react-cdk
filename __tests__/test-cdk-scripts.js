import path from 'path';
import childProcess from 'child_process';
import dirTree from 'directory-tree';
import chalk from 'chalk';


const generateProject = () => new Promise((resolve, reject) => {
  const command = childProcess.spawn('node', ['./scripts/test-cdk-scripts']);

  command.on('close', (code) => {
    if (!code) {
      process.stdout.write(chalk.yellowBright('\nProject is generated successfully!\n\n'));
      resolve(code);
    } else {
      process.stdout.write(chalk.red(`\nTerminated with code: ${code}\n\n`));
      reject(code);
    }
  });

  command.stderr.on('data', (data) => {
    /** Do nothing since could be warning during installation */
  });

  command.stdout.on('data', (data) => {
    process.stdout.write(chalk.green(data));
  });

  process.stdout.write(chalk.yellowBright('\nGenerating a project. It will take some time!...\n\n'));
});

const executeScript = (script) => new Promise((resolve, reject) => {
  const args = script.split(' ');
  const command = childProcess.spawn('yarn', args, {
    cwd: path.resolve(__dirname, '../test-gen/test-cdk-scripts/')
  });

  let stdout = '';
  let stderr = '';
  command.on('close', (code) => {
    if (!code) {
      process.stdout.write(chalk.yellowBright('\nFinished successfully!\n\n'));
      resolve({ code, stdout });
    } else {
      process.stdout.write(chalk.red(`\nTerminated with code: ${code}\n${stderr}\n`));
      reject(code);
    }
  });

  command.stderr.on('data', (data) => {
    /** Do nothing since could be warning during installation */
    stderr = stderr + data;
  });

  command.stdout.on('data', (data) => {
    process.stdout.write(chalk.green(data));
    stdout = stdout + data;
  });

  process.stdout.write(chalk.yellowBright(`\nExecuting script: ${script}. It will take some time!...\n\n`));
});

describe('CDK-Scripts', () => {
  it('Should generate right folder structure by less then 120 sec and run scripts', () => {
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 120000;
    
    expect.assertions(9);
    return generateProject()
      .then(code => {
        /** project should be successfully generated: */

        expect(code).toBe(0);
        return code;
      })
      .then(code => {
        /** project folder should have the correct structure: */

        const projectFolder = dirTree('./test-gen/test-cdk-scripts', { exclude: /node_modules/ });
        expect(projectFolder).toMatchSnapshot('test-gen/test-cdk-scripts');

        /** run prepublish script */
        return executeScript('prepublish');
      })
      .then(({ code, stdout }) => {
        /** script should successfully executed */

        expect(code).toBe(0);
        return stdout;
      })
      .then(stdout => {
        /** dist folder should have the correct structure: */

        const projectFolder = dirTree('./test-gen/test-cdk-scripts/dist', { exclude: /node_modules/ });
        expect(projectFolder).toMatchSnapshot('test-gen/test-cdk-scripts/dist');

        /** run report script */
        return executeScript('cdk-scripts report --info userenv');
      })
      .then(({ code, stdout }) => {
        /** script should successfully executed */

        expect(code).toBe(0);
        return stdout;
      })
      .then(stdout => {
        /** report script should have the correct output */

        expect(stdout).toEqual(expect.stringContaining('Please, include this information to the bug report'));

        /** run welcome script */
        return executeScript('cdk-scripts welcome --show success');
      })
      .then(({ code, stdout }) => {
        /** script should successfully executed */

        expect(code).toBe(0);
        return stdout;
      })
      .then(stdout => {
        /** welcome script should have the correct output */

        expect(stdout).toEqual(expect.stringContaining('For more information visit'));

        return new Promise((resolve, reject) => setTimeout(() => resolve(0), 100))
      })
      .then(code => {
        /** just check that chain is finished */

        expect(code).toBe(0);
      })
      .catch(code => {
        throw(code);
      })
      ;
  });
})

/**
 * @todo: move generateProject() to beforeAll. Split chain to separate tests
 * 
 */
