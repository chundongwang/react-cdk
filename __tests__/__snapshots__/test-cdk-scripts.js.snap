// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-gen/test-cdk-scripts 1`] = `
Object {
  "children": Array [
    Object {
      "extension": "",
      "name": ".babelrc",
      "path": "test-gen/test-cdk-scripts/.babelrc",
      "size": 144,
      "type": "file",
    },
    Object {
      "extension": "",
      "name": ".gitignore",
      "path": "test-gen/test-cdk-scripts/.gitignore",
      "size": 570,
      "type": "file",
    },
    Object {
      "extension": "",
      "name": ".npmignore",
      "path": "test-gen/test-cdk-scripts/.npmignore",
      "size": 28,
      "type": "file",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".js",
          "name": "addons.js",
          "path": "test-gen/test-cdk-scripts/.storybook/addons.js",
          "size": 86,
          "type": "file",
        },
        Object {
          "extension": ".js",
          "name": "config.js",
          "path": "test-gen/test-cdk-scripts/.storybook/config.js",
          "size": 136,
          "type": "file",
        },
      ],
      "name": ".storybook",
      "path": "test-gen/test-cdk-scripts/.storybook",
      "size": 222,
      "type": "directory",
    },
    Object {
      "extension": ".md",
      "name": "README.md",
      "path": "test-gen/test-cdk-scripts/README.md",
      "size": 80,
      "type": "file",
    },
    Object {
      "extension": ".json",
      "name": "package.json",
      "path": "test-gen/test-cdk-scripts/package.json",
      "size": 1071,
      "type": "file",
    },
    Object {
      "children": Array [
        Object {
          "extension": ".css",
          "name": "App.css",
          "path": "test-gen/test-cdk-scripts/src/App.css",
          "size": 377,
          "type": "file",
        },
        Object {
          "extension": ".js",
          "name": "Header.js",
          "path": "test-gen/test-cdk-scripts/src/Header.js",
          "size": 280,
          "type": "file",
        },
        Object {
          "extension": ".js",
          "name": "Header.test.js",
          "path": "test-gen/test-cdk-scripts/src/Header.test.js",
          "size": 235,
          "type": "file",
        },
        Object {
          "extension": ".js",
          "name": "Page.js",
          "path": "test-gen/test-cdk-scripts/src/Page.js",
          "size": 245,
          "type": "file",
        },
        Object {
          "extension": ".js",
          "name": "Page.test.js",
          "path": "test-gen/test-cdk-scripts/src/Page.test.js",
          "size": 232,
          "type": "file",
        },
        Object {
          "extension": ".svg",
          "name": "logo.svg",
          "path": "test-gen/test-cdk-scripts/src/logo.svg",
          "size": 2671,
          "type": "file",
        },
      ],
      "name": "src",
      "path": "test-gen/test-cdk-scripts/src",
      "size": 4040,
      "type": "directory",
    },
    Object {
      "extension": ".lock",
      "name": "yarn.lock",
      "path": "test-gen/test-cdk-scripts/yarn.lock",
      "size": 307787,
      "type": "file",
    },
  ],
  "name": "test-cdk-scripts",
  "path": "./test-gen/test-cdk-scripts",
  "size": 313942,
  "type": "directory",
}
`;

exports[`test-gen/test-cdk-scripts/dist 1`] = `
Object {
  "children": Array [
    Object {
      "extension": ".css",
      "name": "App.css",
      "path": "test-gen/test-cdk-scripts/dist/App.css",
      "size": 377,
      "type": "file",
    },
    Object {
      "extension": ".js",
      "name": "Header.js",
      "path": "test-gen/test-cdk-scripts/dist/Header.js",
      "size": 747,
      "type": "file",
    },
    Object {
      "extension": ".js",
      "name": "Page.js",
      "path": "test-gen/test-cdk-scripts/dist/Page.js",
      "size": 666,
      "type": "file",
    },
    Object {
      "extension": ".svg",
      "name": "logo.svg",
      "path": "test-gen/test-cdk-scripts/dist/logo.svg",
      "size": 2671,
      "type": "file",
    },
  ],
  "name": "dist",
  "path": "./test-gen/test-cdk-scripts/dist",
  "size": 4461,
  "type": "directory",
}
`;
